core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,430 @@
+            } else {
+                requested = LongMath.saturatedAdd(oldRequested, n);
+            }
+
+            if (oldRequested > 0) {
+                // SurroundingSubscriber is publishing data.
+                // New requests will be handled by 'publishDownstream(item)'.
+                return;
+            }
+
+            publish();
+        }
+
+        private void publish() {
+            if (closed || requested <= 0 && upstreamRequested <= 0) {
+                return;
+            }
+
+            switch (state) {
+                case REQUIRE_HEAD: {
+                    sendHead();
+                    break;
+                }
+                case REQUIRE_BODY: {
+                    if (!subscribed) {
+                        subscribed = true;
+                        publisher.subscribe(this, executor, options);
+                        return;
+                    }
+                    if (upstreamRequested > 0) {
+                        return;
+                    }
+                    final Subscription upstream = this.upstream;
+                    if (upstream != null) {
+                       requestUpstream(upstream);
+                    }
+                    break;
+                }
+                case REQUIRE_TAIL: {
+                    sendTail();
+                    break;
+                }
+                case REQUIRE_COMPLETE: {
+                    sendComplete();
+                    break;
+                }
+                case DONE: {
+                    closed = true;
+                    break;
+                }
+            }
+        }
+
+        private void sendHead() {
+            setState(State.REQUIRE_HEAD, State.REQUIRE_BODY);
+            publishDownstream(head);
+        }
+
+        private void sendTail() {
+            setState(State.REQUIRE_TAIL, State.REQUIRE_COMPLETE);
+            if (tail != null) {
+                publishDownstream(tail);
+            } else {
+                sendComplete();
+            }
+        }
+
+        private void sendComplete() {
+            setState(State.REQUIRE_COMPLETE, State.DONE);
+            close0(null);
+        }
+
+        private void requestUpstream(Subscription subscription) {
+            if (requested <= 0) {
+                return;
+            }
+            assert upstreamRequested == 0;
+            upstreamRequested = requested;
+            if (requested < Long.MAX_VALUE) {
+                requested = 0;
+            }
+            subscription.request(upstreamRequested);
+        }
+
+        private void publishDownstream(T item) {
+            requireNonNull(item, "item");
+            if (closed) {
+                return;
+            }
+            downstream.onNext(item);
+            if (0 < upstreamRequested && upstreamRequested < Long.MAX_VALUE) {
+                upstreamRequested--;
+            } else if (requested < Long.MAX_VALUE) {
+                requested--;
+            }
```

Comment:
```suggestion
            }
```
I think we don't need this because `requested` is already decreased to 0 in `requestUpstream()` method

Reply:
```
// elements: head -> 1 -> 2 -> 3 -> tail
request(1): requested: 1, upstreamRequested: 0
  -> onNext(head)

request(3): requested: 0, upstreamRequested: 3
  -> onNext(1, 2, 3)

request(1): requested: 1, upstreamRequested: 0
  -> onNext(tail)
```

oh~ on `head` and `tail` like above case, `requested` is not decreased to 0 cause it doesn't call `requestUpstream()` (on head&tail, we just `sendHead`, `sendTail` directly without `requestUpstream`)

So, on `publishDownstream`, we should decreased `upstreamRequested` for BODY or `requested` for HEAD&TAIL!

Reply:
That's also correct. I forgot the `head` and `tail` at this time also. 😅